tutorials/dataframe/df103_NanoAODHiggsAnalysis.py

```diff
@@ -0,0 +1,574 @@
+## kind) and in a combined plot indicating the decay of the Higgs boson with a
+## mass of about 125 GeV.
+##
+## The following steps are performed for each sample with data and simulated
+## events in order to reconstruct the Higgs boson from the selected muons and
+## electrons:
+## 1. Select interesting events with multiple cuts on event properties, e.g.,
+##    number of leptons, kinematics of the leptons and quality of the tracks.
+## 2. Reconstruct two Z bosons of which only one on the mass shell from the
+##    selected events and apply additional cuts on the reconstructed objects.
+## 3. Reconstruct the Higgs boson from the remaining Z boson candidates and
+##    calculate its invariant mass.
+##
+## Another aim of this version of the tutorial is to show a way to blend C++
+## and Python code. All the functions that make computations on data to define
+## new columns or filter existing ones in a precise way, better suited to be
+## written in C++, have been moved to a header that is then declared to the
+## ROOT C++ interpreter. The functions that instead create nodes of the
+## computational graph (e.g. Filter, Define) remain inside the main Python
+## script.
+##
+## \macro_image
+## \macro_code
+## \macro_output
+##
+## \date July 2019
+## \author Stefan Wunsch (KIT, CERN), Vincenzo Eduardo Padulano (UniMiB, CERN)
+
+import ROOT
+import os
+
+# Enable multi-threading
+ROOT.ROOT.EnableImplicitMT()
+
+# Include necessary header
+tutorial_dir = str(ROOT.gROOT.GetTutorialDir())
+dataframe_dir = "dataframe"
+header_file = "df103_NanoAODHiggsAnalysis_python.h"
+higgs_header_path = os.path.join(
+    os.sep, tutorial_dir + os.sep,
+    dataframe_dir + os.sep, header_file
+)
+
+ROOT.gInterpreter.Declare('#include "{}"'.format(higgs_header_path))
+
+
+# Python functions
+def reco_higgs_to_2el2mu(df):
+    """Reconstruct Higgs from two electrons and two muons"""
+    # Filter interesting events
+    df_base = selection_2el2mu(df)
+    # Compute masses of Z systems
+    df_z_mass = df_base.Define(
+        "Z_mass",
+        "compute_z_masses_2el2mu(Electron_pt, Electron_eta, Electron_phi,"
+        " Electron_mass, Muon_pt, Muon_eta, Muon_phi, Muon_mass)"
+    )
+    # Cut on mass of Z candidates
+    df_z_cut = filter_z_candidates(df_z_mass)
+    # Reconstruct H mass
+    df_h_mass = df_z_cut.Define(
+        "H_mass",
+        "compute_higgs_mass_2el2mu(Electron_pt, Electron_eta, Electron_phi,"
+        " Electron_mass, Muon_pt, Muon_eta, Muon_phi, Muon_mass)"
+    )
+
+    return df_h_mass
+
+
+def selection_2el2mu(df):
+    """Select interesting events with two electrons and two muons"""
+    df_ge2el2mu = df.Filter(
+        "nElectron>=2 && nMuon>=2",
+        "At least two electrons and two muons"
+    )
+    df_eta = df_ge2el2mu.Filter(
+        "All(abs(Electron_eta)<2.5) && All(abs(Muon_eta)<2.4)",
+        "Eta cuts"
+    )
+
+    df_pt = df_eta.Filter("pt_cuts(Muon_pt, Electron_pt)", "Pt cuts")
+
+    df_dr = df_pt.Filter(
```

Comment:
I would put such things in a single line even though you go over 90 characters. But that's just my two cents!

Reply:
Yeah, I agree that the PEP8 coding convention sometimes is a bit too keen on verticalizing the code :( I just tried to be totally compliant with it and I think that generally speaking is better to adhere to a set of rules when there is one already established. I would wait for a third opinion to make a final decision on this

Reply:
You used `yapf` or something similar? However, I would enlarge the number of characters per line to make it more nicely readable (and less lines of code).

Reply:
No, I don't use any automatic formatter (btw I didn't know about `yapf`, thanks!). I just format the code by hand following PEP8 website. In this case for multiline indentation of arguments to functions I follow what's written in [this](https://www.python.org/dev/peps/pep-0008/#id17) paragraph.

Reply:
Ok then, what character limit would you set? PEP8 defines 79 characters per line, maybe 90 would be better? or 120 like ROOT?